includes/Admin/Settings.php

```diff
@@ -7,59 +7,227 @@
+					type="number"
+					name="num_products_to_generate"
+					value="<?php echo esc_attr( self::DEFAULT_NUM_PRODUCTS ); ?>"
+					min="1"
+					<?php disabled( $current_job instanceof AsyncJob ); ?>
+				/>
+				<?php
+				submit_button(
+					'Generate',
+					'primary',
+					'generate_products',
+					false,
+					$generate_button_atts
+				);
+				?>
 			</p>
-			<?php submit_button( 'Generate', 'primary', 'generate_products' ); ?>
 
-			<h2>Generate Orders</h2>
+			<h2>Generate orders</h2>
 			<p>
-				<input type="number" name="num_orders_to_generate" value="<?php echo esc_attr( self::DEFAULT_NUM_ORDERS ); ?>" min="1" />
-				Number of orders to generate.
+				<label for="generate_orders_input" class="screen-reader-text">Number of orders to generate</label>
+				<input
+					id="generate_orders_input"
+					type="number"
+					name="num_orders_to_generate"
+					value="<?php echo esc_attr( self::DEFAULT_NUM_ORDERS ); ?>"
+					min="1"
+					<?php disabled( $current_job instanceof AsyncJob ); ?>
+				/>
+				<?php
+				submit_button(
+					'Generate',
+					'primary',
+					'generate_orders',
+					false,
+					$generate_button_atts
+				);
+				?>
 			</p>
-			<?php submit_button( 'Generate', 'primary', 'generate_orders' ); ?>
 
-			<h2>Cancel all scheduled generations</h2>
-			<p>
-			</p>
-			<?php submit_button( 'Cancel all', 'primary', 'cancel_all_generations' ); ?>
+			<?php
+			submit_button(
+				'Cancel current job',
+				'secondary',
+				'cancel_job',
+				true,
+				$cancel_button_atts
+			);
+			?>
 		</form>
 		<?php
+
+		self::heartbeat_script();
+	}
+
+	/**
+	 * Script to run the progress bar.
+	 *
+	 * @return void
+	 */
+	protected static function heartbeat_script() {
+		?>
+		<script>
+			( function( $ ) {
+				var $progress = $( '#smoothgenerator-progress-bar' );
+				var $controls = $( '[id^="generate_"]' );
+				var $cancel   = $( '#cancel_job' );
+
+				$( document ).on( 'ready', function () {
+					wp.heartbeat.disableSuspend();
+					wp.heartbeat.connectNow();
+					wp.heartbeat.interval( 'fast' );
+				} );
+
+				$( document ).on( 'heartbeat-send', function ( event, data ) {
+					data.smoothgenerator = 'check_async_job_progress';
+				} );
+
+				$( document ).on( 'heartbeat-tick', function ( event, data ) {
+					// Heartbeat and other admin-ajax calls don't trigger wp-cron, so we have to do it manually.
+					$.ajax( {
+						url: data.smoothgenerator_ping_cron,
+						method: 'get',
+						timeout: 5000,
+						dataType: 'html'
+					} );
+
+					if ( 'object' === typeof data.smoothgenerator_async_job_progress ) {
+						var value = parseInt( data.smoothgenerator_async_job_progress.processed );
+						if ( value > 0 ) {
+							$progress.prop( 'value', value );
+						}
+					} else if ( 'complete' === data.smoothgenerator_async_job_progress && $progress.is( ':visible' ) ) {
+						var max = $progress.prop( 'max' );
+						$progress.prop( 'value', max );
+						$progress.parent().append( '✅' );
```

Comment:
Very much a suggestion and not a requirement, but I wonder if we could also change the progress bar label to something like _"Done!"_ at this stage?

Reply:
_Done!_ 3f2386e